content/blog/2019-12-18-heterocl-stream/index.md

```diff
@@ -0,0 +1,158 @@
+
+<img src="group.png" width="300" >
+
+Now, we can start the scheduling process by assigning the timestep to each module. We first perform a very simple as-soon-as-possible (ASAP) algorithm. Namely, the first module within each group will be assigned with timestep 0. After that, we assign the timestep of each module according to the data dependence. An example is shown below.
+
+<img src="schedule1.png" width="300" >
+
+However, this is not correct because as we mentioned above, modules connected with streams should be run in parallel. Namely, they will share the same timestep. To solve that, we add one dotted line back at a time and correct the timesteps. We also need to correct its succeeding modules accordingly. After all dotted lines are added, we finish our scheduling algorithm.
+
+<img src="schedule2.png" width="300" >
+
+Note that there exist cases where we cannot solve. For example, if two modules ``A`` and ``B`` are connected with a solid line, and the producer ``A`` streams to a module ``M`` while ``B`` also streams from ``M``, then there exists no valid scheduling according to our constraints. One possible way to solve that is by merging ``A`` and ``B`` into a new module ``A_B``. In this case, the streaming from/to ``M`` becomes an internal stream, which can be scheduled easily by assigning ``A_B`` and ``C`` with the same timestep.
+
+#### Parallel Execution with Threads
+
+After we assign each module with a timestep, we can start to execute them via threads. Before we execute a module with a new thread, we check whether all modules assigned with smaller timesteps are completed. In other words, we first check whether all modules assigned with smaller timesteps are fired. If not, we schedule the current module to be executed in the future by pushing it into a sorted execution queue. Then, if all modules with smaller timesteps are fired, we check whether they are finished. If not, we perform thread synchronization (e.g., by using ``thread.join()`` in C++). Finally, we need to execute the modules in the execution queue. Since the queue is sorted, we do not need to worry about new modules being inserted into the queue.
+
+#### Stream Buffers
+
+In this work, we implement the streams with buffers that act like FIFOs. Instead of popping from or pushing to the buffers, we maintain a **head** and a **tail** pointer for each buffer. The pointers are stored as integer numbers. The head pointer points to the next element that will be read from, and the tail pointer points to the next element that will be written to. We update the pointers each time an element is written to or read from the buffer. We need to perform modulo operations if the pointer value is greater than the buffer size (i.e., FIFO depth). Since we may have two threads updating the pointers at the same time, we declare them as atomic numbers. Finally, we maintain a map so that we can access a stream according to its ID.

```

Comment:
> we declare them as atomic numbers

This is a little hard to interpret without more context… what exactly is an "atomic number"?

I assume the answer is that you wrote simulation framework in C++, and you used [std:atomic](https://en.cppreference.com/w/cpp/atomic/atomic) as an alternative to using mutexes to synchronize accesses to the counters.